src/jit/emitxarch.cpp

```diff
@@ -7852,16 +7852,19 @@ void emitter::emitDispAddrMode(instrDesc* id, bool noDetail)
     else
     {
         // Munge any pointers if we want diff-able disassembly
-        if (emitComp->opts.disDiffable)
+        if (!frameRef && emitComp->opts.disDiffable && (static_cast<size_t>((disp >> 20) + 1) > 1))
         {
-            ssize_t top12bits = (disp >> 20);
-            if ((top12bits != 0) && (top12bits != -1))
+            if ((disp < 0) && (disp >= -0xFFFFFF))
```

Comment:
This seems somewhat inconsistent in addition to changing the pre-existing behavior. I understand this corresponds to the "else" case after the check <code>((disp & 0x7F000000) != 0x7F000000)</code> previously at line 7901 but it's somewhat out of tune with the previous "top12bits" check as that looks at the bits 20 and higher while the 0xFFFFFF check looks at bits 24+.

My initial understanding of the rationale for the "top12bits" check is something like, "OK, if [the absolute value of] the offset is big, it's probably not an actual offset but a true address after all". This is consistent with your modified logic that ends up printing "D1FFAB1EH" at line 7865. This secondary disp check however means that values between -16M and -1M will get printed as "-D1FFAB1EH" instead of "+D1FFAB1EH".

I don't understand why we should care for the sign if we come to the conclusion that the disp is actually an RVA, not an offset; also please note that the previous logic used the check for low-order 24 bits (the 0x7F000000 check) to make sure that "values between -16M and 0 are printed as negative offsets".

Your logic changes them to be printed as "-D1FFAB1EH" (in the diffable case), and, if I'm right to understand that the entire purpose of the "D1FFAB1EH" transformation is to get rid of position-dependent constants in the disassembly, they have the potential to cause spurious diffs with offsets that happen to end up resolving to "+D1FFAB1EH".

I don't think this is a blocking issue as I suspect that a disassembly diff obtained in this manner can still exhibit somewhat noisy behavior w.r.t. "disp" transformations - for instance, we're never going to be able to "munge any pointers" for a small executable with size under 1MB - but it would be nice to understand whether there's some deeper reason for the 1MB and 16MB constant choices or whether they're just arbitrary heuristic constants that happened to diverge over time or due to an overlooking.

Reply:
I agree with Tomas.

Reply:
Fixed the "-D1FFAB1EH" printing so now we always print "D1FFAB1EH" when disp is outside of the range (-1M, +1M) and `disDiffable` is enabled - this part looks OK to me.

The `((disp & 0x7F000000) != 0x7F000000)` check is somewhat confusing for me as well, but it might not be directly related to diffable disasm - it's more likely to be hit when `disDiffable` is disabled and in that scenario the offset vs pointer detection can be more relaxed. In any case, I only changed the condition so that it's more clear that a range check is being done (`ssize_t` could be 64-bit, but I believe the constants can't be 64-bit) and I added the missing `nsep` check there.

Reply:
The only previous change related to this part of code was in #19040 and from there also it looks like there is no correlation intended between these two checks.

Reply:
LGTM, thanks for addressing my feedback!